### PR TITLE
Docker: development & production profiles

### PR DESCRIPTION
**Pull request by delmaass**
Containers :
- db : PostgreSQL database for the backend
- backend : Django server (port 8100)
- frontend-dev : CRA dev environment (port 3100)
- frontend-prod : Built with Yarn then served with Nginx (80 -> port 3100)

The frontend host should be set in the environment variable of Django to be allowed.

The PR won't pass Blackend and Prettier CI since there are issues with them (see #18).

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    environment:
+      - CHOKIDAR_USEPOLLING=true
+    profiles: ["dev"]
+
+  frontend-prod:
+    build:
+      context: frontend
+      dockerfile: prod.Dockerfile
+    ports:
+      - '3100:80'
```
Comment (abocquet) [inline docker-compose.yml:55]: I'm not sure if we should use port `80` here for now. There is a good chance one might use nginx as a proxy on top of the whole app. Let's use port 3000 as for frontend-dev for the moment

File: frontend/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+WORKDIR /app
+
+# add `/app/node_modules/.bin` to $PATH
+ENV PATH /app/node_modules/.bin:$PATH
+
+# install app dependencies
+COPY package.json ./
+COPY package-lock.json ./
+RUN npm install --silent
+RUN npm install react-scripts@3.4.1 -g --silent
```
Comment (abocquet) [inline frontend/Dockerfile:14]: What does `react-scripts` do?
Why don't we add it in package.json?

File: frontend/nginx/nginx.conf [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    try_files $uri $uri/ /index.html;
+  }
+
+  error_page   500 502 503 504  /50x.html;
+
+  location = /50x.html {
+    root   /usr/share/nginx/html;
+  }
+
+}
```
Comment (abocquet) [inline frontend/nginx/nginx.conf:17]: Good call on nginx here, maybe we should discuss for another one in front of frontend + backend so as to secure both AND the file server
It'll be also useful to handle SSL/HTTPS on the whole app level

File: frontend/prod.Dockerfile [hunk 4]
```diff
@@ -0,0 +1,18 @@
+# build environment
+FROM node:13.12.0-alpine as build
+WORKDIR /app
+ENV PATH /app/node_modules/.bin:$PATH
+COPY package.json ./
+COPY package-lock.json ./
+RUN npm ci --silent
+RUN npm install react-scripts@3.4.1 -g --silent
```
Comment (abocquet) [inline frontend/prod.Dockerfile:8]: Same comment as before

File: frontend/prod.Dockerfile [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+COPY . ./
+RUN npm run build
+
+# production environment
+FROM nginx:stable-alpine
+COPY --from=build /app/build /usr/share/nginx/html
+# React router
+COPY nginx/nginx.conf /etc/nginx/conf.d/default.conf
+EXPOSE 80
```
Comment (abocquet) [inline frontend/prod.Dockerfile:17]: Same comment as before (port 80->3000)

File: sso/settings.py [hunk 6]
```diff
@@ -31,7 +31,7 @@
 # SECURITY WARNING: don't run with debug turned on in production!
 DEBUG = env.bool("DEBUG")
 
-ALLOWED_HOSTS = []
+ALLOWED_HOSTS = ["localhost"] + [env.str("FRONTEND_HOST")] + ["127.0.0.1"]
```
Comment (abocquet) [inline sso/settings.py:34]:
```suggestion
ALLOWED_HOSTS = ["localhost", "127.0.0.1"] + [env.str("FRONTEND_HOST")]
```

File: sso/settings.py [hunk 7]
```diff
@@ -59,7 +59,7 @@
     "django.middleware.clickjacking.XFrameOptionsMiddleware",
 ]
 
-CORS_ORIGIN_WHITELIST = ("http://localhost:3000",)
+CORS_ORIGIN_WHITELIST = ("http://localhost:3100",)
```
Comment (abocquet) [inline sso/settings.py:62]: I'm not sure this is going to work in prod from a distant web browser, but it's not crucial for the moment

**Comment by abocquet** [comment]
This is a good PR, but I'm not approving it for the moment, as we should take a moment to think about how this will be deployed in prod

File: sso/settings.py [hunk 8]
```diff
@@ -57,7 +57,16 @@
...
-CORS_ORIGIN_WHITELIST = ("http://localhost:3100",)
+PRODUCTION_URL = (
+    env.str("PRODUCTION_SUBDOMAIN") + "." + env.str("PRODUCTION_DOMAIN")
+    if env.str("PRODUCTION_SUBDOMAIN")
+    else env.str("PRODUCTION_DOMAIN")
+)
+CORS_ORIGIN_WHITELIST = (
+    "http://localhost:3001",
+    "http://127.0.0.1:3001",
+    "http://" + PRODUCTION_URL,
```
Comment (abocquet) [inline sso/settings.py:68]:
```suggestion
    "https://" + PRODUCTION_URL,
```

**Review by abocquet — APPROVED** [review]

**Comment by delmaass** [comment]
Hi @abocquet ! Thanks for your last review :) Can you give this PR a last review please ? This docker configuration should match with the architecture described in the RFC for the SSO app.

**Pull request merged**